src/benchmark/file_io_read_micro_benchmark.cpp

```diff
@@ -0,0 +1,205 @@
+    if (write(fd, std::data(numbers), BUFFER_SIZE_MB * MB) != BUFFER_SIZE_MB * MB) {
+      std::cout << "write error" << std::endl;
+    }
+    close(fd);
+  }
+
+  void TearDown(::benchmark::State& /*state*/) override {
+    // TODO(everybody): Error handling
+    std::remove("file.txt");
+  }
+
+ protected:
+};
+
+BENCHMARK_DEFINE_F(FileIOMicroReadBenchmarkFixture, READ_NON_ATOMIC)(benchmark::State& state) {  // open file
+  int32_t fd;
+  if ((fd = open("file.txt", O_RDONLY)) < 0) {
+    std::cout << "open error " << errno << std::endl;
+  }
+  const int32_t NUMBER_OF_BYTES = state.range(0) * MB;
+
+  for (auto _ : state) {
+    state.PauseTiming();
+    micro_benchmark_clear_disk_cache();
+    std::vector<uint32_t> read_data;
+
+    read_data.resize(NUMBER_OF_BYTES / 4);
+    state.ResumeTiming();
+
+    lseek(fd, 0, SEEK_SET);
+
+    if (read(fd, std::data(read_data), NUMBER_OF_BYTES) != NUMBER_OF_BYTES) {
+      Fail("read error: " + strerror(errno));
+    }
+    state.PauseTiming();
+    auto sum = std::accumulate(read_data.begin(), read_data.end(), uint64_t{0});
+    // sum == 0 because read vector is empty
+    Assert(control_sum == sum, "Sanity check failed: Not the same result");
+    state.ResumeTiming();
+  }
+}
+
+BENCHMARK_DEFINE_F(FileIOMicroReadBenchmarkFixture, PREAD_ATOMIC)(benchmark::State& state) {
+  int32_t fd;
+  if ((fd = open("file.txt", O_RDONLY)) < 0) {
+    std::cout << "open error " << errno << std::endl;
+  }
+  const int32_t NUMBER_OF_BYTES = state.range(0) * MB;
+
+  for (auto _ : state) {
+    state.PauseTiming();
+    micro_benchmark_clear_disk_cache();
+    std::vector<uint32_t> read_data;
+    read_data.resize(NUMBER_OF_BYTES / 4);
+    state.ResumeTiming();
+
+    if (pread(fd, std::data(read_data), NUMBER_OF_BYTES, 0) != NUMBER_OF_BYTES) {
+      Fail("read error: " + strerror(errno));
+    }
+
+    state.PauseTiming();
+    auto sum = std::accumulate(read_data.begin(), read_data.end(), uint64_t{0});
+    Assert(control_sum == sum, "Sanity check failed: Not the same result");
+    state.ResumeTiming();
+  }
+}
+
+BENCHMARK_DEFINE_F(FileIOMicroReadBenchmarkFixture, MMAP_ATOMIC_MAP_PRIVATE)(benchmark::State& state) {
+  int32_t fd;
+  if ((fd = open("file.txt", O_RDONLY)) < 0) {
+    std::cout << "open error " << errno << std::endl;
+  }
+  const uint32_t NUMBER_OF_BYTES = state.range(0) * MB;
+
+  for (auto _ : state) {
+    state.PauseTiming();
+    micro_benchmark_clear_disk_cache();
+    std::vector<uint32_t> read_data;
+    read_data.resize(NUMBER_OF_BYTES / 4);
+    state.ResumeTiming();
+
+    // Getting the mapping to memory.
+    off_t OFFSET = 0;
+
+    int32_t* map = reinterpret_cast<int32_t*>(mmap(NULL, NUMBER_OF_BYTES, PROT_READ, MAP_PRIVATE, fd, OFFSET));
+    if (map == MAP_FAILED) {
+      std::cout << "Mapping Failed. " << std::strerror(errno) << std::endl;
+      continue;
+    }
+
+    memcpy(std::data(read_data), map, NUMBER_OF_BYTES);
```

Comment:
Copying the entire data set defeats the purpose of using mmap. Even though you cannot directly lay an vector of the mapped data, please try to build your check sum on the mmap'ed file.